nifi-nar-bundles/nifi-jms-bundle/nifi-jms-processors/src/main/java/org/apache/nifi/jms/processors/strategy/consumer/record/RecordWriter.java

```diff
@@ -0,0 +1,241 @@
+                        OutputStrategy outputStrategy,
+                        ComponentLog logger) {
+        this.readerFactory = readerFactory;
+        this.writerFactory = writerFactory;
+        this.serializer = serializer;
+        this.outputStrategy = outputStrategy;
+        this.logger = logger;
+    }
+
+    @Override
+    public void write(ProcessSession session, List<T> messages, AttributeSupplier<T> attributeSupplier, MessageConsumerCallback<T> messageConsumerCallback) {
+        consumeMessagesIntoRecords(session, messages, attributeSupplier, messageConsumerCallback);
+    }
+
+    private void consumeMessagesIntoRecords(ProcessSession session, List<T> messages, AttributeSupplier<T> attributeSupplier, MessageConsumerCallback<T> messageConsumerCallback) {
+        FlowFile flowFile = session.create();
+
+        final Map<String, String> attributes = new HashMap<>();
+        final AtomicInteger recordCount = new AtomicInteger();
+
+        final List<T> processedMessages = new ArrayList<>();
+        final List<T> failedMessages = new ArrayList<>();
+
+        RecordSetWriter writer = null;
+        boolean isWriterInitialized = false;
+
+        try {
+            for (T message : messages) {
+                if (message == null) {
+                    break;
+                }
+
+                final byte[] recordBytes = serializer.serialize(message);
+                try (final InputStream in = new ByteArrayInputStream(recordBytes)) {
+                    final RecordReader reader;
+
+                    // parse incoming message which may contain multiple messages
+                    try {
+                        reader = readerFactory.createRecordReader(attributes, in, recordBytes.length, logger);
+                    } catch (final IOException ioe) {
+                        logger.error("Failed to parse message due to comms failure. Will roll back session and try again momentarily.");
+                        messageConsumerCallback.onFailure(flowFile, processedMessages, failedMessages, ioe);
+                        closeWriter(writer);
+                        return;
+                    } catch (final Exception e) {
+                        logger.error("Failed to parse message, sending to the parse failure relationship", e);
+                        failedMessages.add(message);
+                        messageConsumerCallback.onParseFailure(flowFile, message, e);
+                        continue;
+                    }
+
+                    // write messages as records into FlowFile
+                    try {
+                        Record record;
+                        while ((record = reader.nextRecord()) != null) {
+
+                            if (attributeSupplier != null && !outputStrategy.equals(USE_VALUE)) {
+                                final Map<String, String> additionalAttributes = attributeSupplier.supply(message);
+                                if (outputStrategy.equals(USE_APPENDER)) {
```

Comment:
Having such strategies would be good. I like the idea. However, it is only useful if those strategies' signatures are the same. But unfortunately, this is not the case. AppendingOutputStrategy doesn't need the originalRecordKey. I can "artificially" add it, we could say that we even make append smarter because you can prefix the original records too, but it would be against the concept of append. I also don't think that it would be good performance-wise.